Update FilePath constructors to allow a non-partial way to create a FilePath

rhagenson opened the pull request:
Closes #3820

As part of [RFC 70](https://github.com/ponylang/rfcs/blob/main/text/0070-filepath-constructor.md), `FilePath` now has two constructors: `create` and `from`. Where previously `create` was a partial constructor and thus could fail if called with a `FilePath` base, now `create` is a complete constructor and will always succeed. Partial construction with `FilePath` now exists in the `from` constructor.

A new package-level root authority, `FileAuth`, has been added to allow explicit "files" authority without `AmbientAuth`. This new authority follows the same convention as is used in the "net" package -- note: this change is different from the the accepted RFC in order to maintain consistency within stdlib 's use of package-level root authority. Any prior instance where `AmbientAuth` was accepted in relation to `FilePath` now also accepts `FileAuth` with the same result.

Additional tests have been added to cover the expected error behavior of `FileAuth.from(...)`, as well as the creation and use of `FileAuth` in place of `AmbientAuth`.

Theodus commented:
Closes #3820

SeanTAllen commented:
I'm loiking forward to this one.

rhagenson commented:
@SeanTAllen This is (I believe) my first contribution to the stdlib so I am looking forward to it as well. 

I have all the tests passing now. I still need to add the additional test(s) for the change and want to look at stdlib/tests to see if I can reduce the size of try-blocks now that `create` cannot `error`. The former is guaranteed, the latter is "with this change can we have cleaner code in stdlib" thinking.

Also kind of surprised the following in `mkdtemp` works:

```pony
...
    let parent: FilePath = match base
      | let base': FileAuth => FilePath(base', dir)
      | let base': FilePath => FilePath.from(base', dir)?
    end
...
```

I would expect `parent: (FilePath | None)` with `FilePath.from(base', dir)?` being partial. Or, alternatively `FilePath.from(base', dir)?` to require a try-block.

SeanTAllen commented:
`mkdtemp` is defined as partial. I assume that hasn't changed...

```pony
  new val mkdtemp(
    base: (FilePath | AmbientAuth),
    prefix: String = "",
    caps': FileCaps val = recover val FileCaps .> all() end)
    ?
```

and it would explain the working without a try.

rhagenson commented:
Ah, I was not thinking big enough. I had assumed the part I rewrote to now only sometimes failing would have bubbled up more, but at the function level it is already denoted as possibly failing.

rhagenson commented:
I finished the split of FilePath construction to be `create` with new `FileAuth` and `from` with `FilePath`, fixed existing code in stdlib for this change, added tests for successful and failing serial construction. Last thing I know I need to write for this is a changelog entry, but I am going to open it up for review prior to sync tomorrow.

SeanTAllen commented on packages/files/_test.pony:
```diff
@@ -189,10 +194,37 @@ class iso _TestSymlink is UnitTest
+      let path = "tmp.filepath"
+      let ambient = h.env.root as AmbientAuth
+      let filepath = FilePath(FileAuth(ambient), path)
+      h.assert_error({()? => FilePath.from(filepath, "/")? })
```
it took me a minute to figure out why this should be an error.  i think a comment on the test class explaining why an error is expected would be good. i think the same but in reverse could be said for the other test.

also as part of that I think these could have better names.

something like,

"test file path from errors when X" 

and

"test file path from doesn't error when Y"

the current names say what the expected outcome is, not what is being tested.

SeanTAllen commented on packages/files/_test.pony:
```diff
@@ -189,10 +194,37 @@ class iso _TestSymlink is UnitTest
+      let filepath = FilePath(FileAuth(ambient), path)
+      h.assert_error({()? => FilePath.from(filepath, "/")? })
+    else
+      h.fail("Unhandled error!")
```
when does this happen? if env.root isn't ambient? does assert_error cover the other error?

if yes, I think this should have a more informative error failure message.

SeanTAllen commented on packages/files/_test.pony:
```diff
@@ -189,10 +194,37 @@ class iso _TestSymlink is UnitTest
+      let filepath = FilePath(FileAuth(ambient), path)
+      h.assert_no_error({()? => FilePath.from(filepath, path)? })
+    else
+      h.fail("Unhandled error!")
```
i think a more informative fail message would be good here. ok, we got an error, why might that be?

SeanTAllen commented on packages/files/_test.pony:
```diff
@@ -189,10 +194,37 @@ class iso _TestSymlink is UnitTest
+  fun apply(h: TestHelper) =>
+    try
+      let path = "tmp.filepath"
+      let ambient = h.env.root as AmbientAuth
```
both this and the other test I think would be better as

```pony
match h.env.root
| let auth: Ambient Auth =>
    ... logic...
else
    h.fail("env.root isn't AmbientAuth.")
end
```

as they are currently constructed, the `as` is within the `try` block which makes the cause of an error unpredictable.
i think the only thing that can error within the `try` should be the `FilePath.from` call.

SeanTAllen commented on packages/files/_test.pony:
```diff
@@ -189,10 +194,37 @@ class iso _TestSymlink is UnitTest
+      h.fail("Unhandled error!")
+    end
+
+
```
should only be 1 blank line between test classes, you can kill this.

SeanTAllen commented on packages/files/file_path.pony:
```diff
@@ -54,43 +54,50 @@ class val FilePath
+    base: FilePath,
+    path': String,
+    caps': FileCaps val = recover val FileCaps .> all() end)
+    ?
```
per style guide, the `?` should be on line 74 at the end

SeanTAllen commented on .release-notes/split-filepath.md:
```diff
@@ -0,0 +1,3 @@
+## Split `FilePath` construction into two methods
+
+`FilePath` previously had only one constructor, the default `create`, which used a union of `(AmbientAuth | FilePath)`. The first half of this union `AmbientAuth` could never `error`, while the second `FilePath` could `error`. By splitting construction into two methods, we now have a default `create` which cannot `error` -- now using a "files" package root authority `FileAuth` rather than global  ...
```
the breaking change to the `mkdtemp` constructor should also be noted and covered with a before and after.

SeanTAllen commented:
@rhagenson before this gets merged, can you update the first comment on this PR to be a good summation of the changes that this encompasses and link to the rfc? We'll use that as the squashed commit comment and if anyone follows the link from the CHANGELOG to this PR, they will find a good summary at the top.

SeanTAllen commented on packages/files/file_path.pony:
```diff
@@ -54,43 +54,50 @@ class val FilePath
 
   new val create(
-    base: (FilePath | AmbientAuth),
+    base: FileAuth,
```
So I believe this was supposed to be like TCPConnectionAuth which means this should take FileAuth | AmbientAuth.

`TCPConnectionAuth` is type `TCPConnectionAuth is (AmbientAuth | NetAuth | TCPAuth | TCPConnectAuth)`

If this takes `FileAuth | AmbientAuth` then a lot of code that worked previously will still work. IE its less of a breaking change and a ton of the example and test changes you had to do don't have to happen.

I think it makes sense to have a `FilePathAuth` that is `type FilePathAuth is (AmbientAuth | FileAuth)` and make this `FilePathAuth` rather than `FileAuth`.

rhagenson replied:
I had the definition of the new `create` in the RFC -- the primary reason for the creation of a "files" root authority was avoiding judicious use of `AmbientAuth`.

SeanTAllen replied:
I didnt read the RFC close enough. I saw "like TCPConnection". You can leave as is, but then I'll open an RFC to add AmbientAuth back in.

SeanTAllen commented:
@rhagenson let me know if you would like any assistance with this.

rhagenson commented:
I will make a point to get all the suggested changes in place today. I have been wrapped up in other matters so have not gotten back to this. I recognize that y'all are waiting on me so moving it up in priority. I do not think I will need assistance.

SeanTAllen commented:
@rhagenson no rush. was checking, nothing more.

rhagenson commented on examples/mandelbrot/mandelbrot.pony:
```diff
@@ -73,7 +73,7 @@ class val Config
 
   new val create(env: Env) ? =>
-    let cs = CommandSpec.leaf("gups_opt",
+    let cs = CommandSpec.leaf("run",
```
The mandelbrot example needed to be fixed and I noticed in the process that the mandelbrot example had been copied from the gups_opt example without changing the command accordingly. No need to include this, but a change I figured as minor enough that it could be cleaned up in the process. Let me know if you want it removed.

SeanTAllen replied:
I think at this point as long as it is noted in the PR that is fine.

rhagenson commented on packages/files/_test.pony:
```diff
@@ -18,6 +18,8 @@ actor Main is TestList
     test(_TestWalk)
+    test(_TestFilePathFrom)
+    test(_TestFilePathFromError)
     test(_TestDirectoryOpen)
     test(_TestDirectoryFileOpen)
     test(_TestPathClean)
```
Do we want to rope that into this PR? Minor task so nothing that would take too long. Biggest hurdle is how we want to sort them. My suggested order is alphabetical across all systems, blank line, then system specific/exclusion tests with comments above the two "sections" of tests. By this I also mean that there would be one block for each system specific/exclusion set of tests.

SeanTAllen replied:
I think opening an issue and doing as a different PR is best.  Otherwise its a lot of noise in this PR distracting from the changes.

rhagenson commented on examples/mandelbrot/mandelbrot.pony:
```diff
@@ -114,7 +114,8 @@ class val Config
     outpath =
       try
-        FilePath(env.root as AmbientAuth, cmd.option("output").string())?
+        let auth = env.root as AmbientAuth
+        FilePath(FileAuth(auth), cmd.option("output").string())
       else
```
I have no issue making this change. I had it the other way around to avoid a long line and as you said, it it not really that long of a line anyhow.

rhagenson commented:
@SeanTAllen Top comment rewritten and all direct mentions of `AmbientAuth` back to using `AmbientAuth` (i.e., revert my `FileAuth(ambient)` wrappings). 

Please let me know if there is anything else that is needed.

ergl commented on packages/files/file_path.pony:
```diff
@@ -54,70 +54,78 @@ class val FilePath
 
-    The resulting FilePath will have capabilities that are the intersection of
-    the supplied capabilities and the capabilities on the parent.
+    Capabilities are as provided, unlike with `FilePath.from(...)`.
```
I'd probably reword this a bit, perhaps by moving it to the first line of the documentation. For example: "Create a new path with the given capabilities", or similar.

rhagenson replied:
I did try to clarify via rewording and expanding, but I did not reorder the docstring as it follows the order of arguments. First line is with respect to base, second is about path, and the third is about caps. Once I push the changes up please take a look and let me know it is still reads poorly. I am not against reordering per se, but trying to keep docstrings of create and from consistent.

ergl commented on packages/files/file_path.pony:
```diff
@@ -54,70 +54,78 @@ class val FilePath
 
-    If AmbientAuth is provided, pattern will be relative to the program's
+    If `AmbientAuth` or `FileAuth` are provided, pattern will be relative to the program's
     working directory. Otherwise, it will be relative to the existing
-    FilePath, and the existing FilePath must be a prefix of the resulting path.
+    `FilePath`, and the existing `FilePath` must be a prefix of the resulting path.
```
I was confused at the mention of the  word `pattern` here, since it's not defined anywhere else. This comes from the [original implementation of `mkdtemp`](https://github.com/ponylang/ponyc/commit/346b5bb20edbc6b90873ad7d8c33585493f846ab), which took a parameter `pattern` instead of the current `prefix`.

I'd probably reword this and say "If `AmbientAuth` or `FileAuth` are provided, the resulting `Filepath` will be relative ....`

We can either do this here or in a separate PR, I don't mind.

rhagenson replied:
I will get it fixed now.

ergl commented on .release-notes/split-filepath.md:
```diff
@@ -0,0 +1,63 @@
+
+The result of this change is that three user changes are needed, namely around `create`, `from`, and `mkdtemp`. Any place where previously `AmbientAuth` was accepted now also accepts `FileAuth`.
+
+Before `create` using `AmbientAuth` to now `create` using `AmbientAuth` or `FileAuth` -- note can no longer fail:
```
These lines and the equivalents below read a bit weird to me, perhaps adding some punctuation would help?

ergl reviewed (commented):
Looks good! I have left some comments below about the documentation blocks

SeanTAllen commented on examples/files/files.pony:
```diff
@@ -5,8 +5,9 @@ actor Main
 
     try
+      let auth = env.root as AmbientAuth
       with file = OpenFile(
-        FilePath(env.root as AmbientAuth, env.args(1)?, caps)?) as File
+        FilePath(auth, env.args(1)?, caps)) as File
```
I think this change with let auth temporary variable should be reverted. it hides the "actual change".

SeanTAllen commented:
This looks good to me. I added one small change request and otherwise, I do agree with @ergl's couple of notes. Looking pretty good though.

rhagenson commented:
@ergl and @SeanTAllen I took at your comments into account. While fixing the removal of `let auth = env.root as AmbientAuth` in Sean's comment I did another search for instances of my previously necessary `FileAuth(auth)` wrapping and removed more of them -- namely in `process/_test.pony`.
